// @sap-ux/fiori-annotation-api/dist/annotation-provider.js
"use strict";
/* eslint-disable @typescript-eslint/consistent-type-imports */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CdsAnnotationProvider = void 0;
exports.getXmlServiceArtifacts = getXmlServiceArtifacts;
exports.uniformUrl = uniformUrl;
const project_access_1 = require("@sap-ux/project-access");
const odata_vocabularies_1 = require("@sap-ux/odata-vocabularies");
const ux_cds_compiler_facade_1 = require("@sap/ux-cds-compiler-facade");
const odata_entity_model_1 = require("@sap-ux/odata-entity-model");
const xml_1 = require("./xml");
const odata_annotation_core_1 = require("@sap-ux/odata-annotation-core");
const vocabularies_1 = require("./vocabularies");
const node_url_1 = require("node:url");
const cds_odata_annotation_converter_1 = require("@sap-ux/cds-odata-annotation-converter");
/**
 * Get XML service artifacts.
 *
 * @param odataVersion - Service OData version.
 * @param path - Service path.
 * @param metadataFile - Metadata file.
 * @param annotationFiles - Annotation files.
 * @param fileCache - File cache.
 * @returns Service artifacts.
 */
function getXmlServiceArtifacts(odataVersion, path, metadataFile, annotationFiles, fileCache) {
    const adapter = new xml_1.XMLAnnotationServiceAdapter({
        type: 'local-edmx',
        odataVersion: odataVersion,
        metadataFile,
        annotationFiles
    }, xml_1.XML_VOCABULARY_SERVICE, { apps: {}, projectType: 'EDMXBackend', root: '' }, '');
    adapter.sync(fileCache);
    const documents = adapter.getDocuments();
    const aliasInformation = getAliasInfo(adapter.metadataService, xml_1.XML_VOCABULARY_SERVICE, documents);
    return {
        path,
        metadataService: adapter.metadataService,
        annotationFiles: documents,
        aliasInfo: aliasInformation,
        fileSequence: adapter.getAllFiles().map((file) => file.uri)
    };
}
/**
 * Normalizes a URL by replacing backslashes with forward slashes and removing leading slashes.
 *
 * @param url - The URL to normalize.
 * @returns The normalized URL.
 */
function uniformUrl(url) {
    return url
        .replaceAll('\\', '/')
        .replaceAll('//', '/')
        .replaceAll(/(?:^\/)/g, '');
}
/**
 *
 */
class CdsAnnotationProvider {
    static serviceInfoCache = new Map();
    static serviceArtifactCache = new Map();
    static cdsCache = new Map();
    static vocabularyService = new odata_vocabularies_1.VocabularyService(true);
    /**
     * Get CDS service artifacts.
     *
     * @param rootPath - Project root path.
     * @param servicePath - Service path.
     * @param fileCache - File cache.
     * @returns Service artifacts or undefined.
     */
    static getCdsServiceArtifacts(rootPath, servicePath, fileCache) {
        let cachedArtifactsByRoot = this.serviceArtifactCache.get(rootPath);
        if (cachedArtifactsByRoot) {
            const cachedService = cachedArtifactsByRoot[servicePath];
            if (cachedService) {
                return cachedService;
            }
        }
        const facade = this.getFacade(rootPath, fileCache, false);
        const services = this.serviceInfoCache.get(rootPath) ?? [];
        const serviceInfo = services.find((s) => uniformUrl(s.urlPath) === uniformUrl(servicePath));
        if (!serviceInfo) {
            return undefined;
        }
        const serviceName = serviceInfo.name;
        const annotationFiles = {};
        const metadataElementMap = facade.getMetadata(serviceName);
        // We collect already full metadata from compile model, we don't need to build it based on paths.
        const metadataCollector = (0, ux_cds_compiler_facade_1.createMetadataCollector)(new Map(), facade);
        const { propagationMap } = facade.getPropagatedTargetMap(serviceName);
        for (const path of facade.getAllSourceUris()) {
            const uri = (0, node_url_1.pathToFileURL)(path).toString();
            const cdsAnnotationFile = (0, cds_odata_annotation_converter_1.toTargetMap)(facade.blitzIndex.forUri(uri), uri, this.vocabularyService, facade);
            const annotationFile = (0, cds_odata_annotation_converter_1.toAnnotationFile)(uri, this.vocabularyService, cdsAnnotationFile, metadataCollector, undefined, propagationMap, true).file;
            annotationFiles[uri] = annotationFile;
        }
        const metadataElements = (0, ux_cds_compiler_facade_1.getMetadataElementsFromMap)(metadataElementMap);
        const metadataService = new odata_entity_model_1.MetadataService({ uriMap: facade?.getUriMap() || new Map() });
        metadataService.import(metadataElements, 'DummyMetadataFileUri');
        const aliasInformation = getAliasInfo(metadataService, xml_1.XML_VOCABULARY_SERVICE, annotationFiles);
        const url = uniformUrl(serviceInfo.urlPath);
        const artifacts = {
            path: url,
            metadataService,
            annotationFiles,
            aliasInfo: aliasInformation,
            fileSequence: facade.getFileSequence().map((path) => (0, node_url_1.pathToFileURL)(path).toString())
        };
        cachedArtifactsByRoot = {};
        cachedArtifactsByRoot[url] = artifacts;
        this.serviceArtifactCache.set(rootPath, cachedArtifactsByRoot);
        return artifacts;
    }
    /**
     * Reset CDS cache clearing compile model and service artifacts.
     *
     * @param rootPath - Project root path.
     * @param fileCache - File cache.
     */
    static resetCache(rootPath, fileCache) {
        this.serviceArtifactCache.delete(rootPath);
        this.getFacade(rootPath, fileCache, true);
    }
    static getFacade(rootPath, fileCache, ignoreCache = false) {
        const cachedValue = this.cdsCache.get(rootPath);
        if (cachedValue && ignoreCache === false) {
            return cachedValue;
        }
        const cache = new Proxy({}, {
            get(compilerCache, path) {
                const cachedValue = compilerCache[path];
                if (cachedValue !== undefined) {
                    return cachedValue;
                }
                const uri = (0, node_url_1.pathToFileURL)(path).toString();
                const value = fileCache.get(uri);
                compilerCache[path] = value;
                return value;
            }
        });
        const facade = (0, ux_cds_compiler_facade_1.createCdsCompilerFacadeForRootSync)(rootPath, [], cache);
        const services = (0, project_access_1.processServices)(facade.getServiceInfo());
        this.serviceInfoCache.set(rootPath, services);
        this.cdsCache.set(rootPath, facade);
        return facade;
    }
}
exports.CdsAnnotationProvider = CdsAnnotationProvider;
function getAliasInfo(metadataService, vocabularyService, files) {
    const aliasInformation = {};
    for (const [uri, document] of Object.entries(files)) {
        const namespaces = (0, odata_annotation_core_1.getAllNamespacesAndReferences)(document.namespace ?? { name: '', type: 'namespace' }, document.references);
        const aliasInfo = (0, odata_annotation_core_1.getAliasInformation)(namespaces, metadataService.getNamespaces());
        const aliasInfoWithAllVocabularies = (0, vocabularies_1.addAllVocabulariesToAliasInformation)(aliasInfo, vocabularyService.getVocabularies());
        aliasInformation[uri] = aliasInfoWithAllVocabularies;
    }
    return aliasInformation;
}
